### PR TITLE
[Enhance] Sampling points based on distance metric

### PR DESCRIPTION
**Pull request by wHao-Wu**
* Sample points based on distance metric (keeping all the far points and sampling only the near points).
* Rename function IndoorPointSample to PointSample (with backward compatibility).
* Support `new_points = points[4:12, 2]` with single attribute in which points are `BasePoints` object.

### CODE REVIEW

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (ZwwWayne) [inline mmdet3d/datasets/pipelines/transforms_3d.py:940]: We do not need to override __call__.

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (ZwwWayne) [inline mmdet3d/datasets/pipelines/transforms_3d.py:934]: should use `__init__(self, *args, **kwargs):` and directly pass it to the super class.

File: mmdet3d/core/points/base_points.py
Comment (ZwwWayne) [inline mmdet3d/core/points/base_points.py:308]: Should add unit tests for this.

**Comment by ZwwWayne** [comment]
Also update configs.

**Comment by wHao-Wu** [comment]
@ZwwWayne 
Whether we need to update the BC-breaking doc [here](https://github.com/open-mmlab/mmdetection3d/blob/master/docs/compatibility.md)?

**Comment by ZwwWayne** [comment]
> @ZwwWayne
> Whether we need to update the BC-breaking doc [here](https://github.com/open-mmlab/mmdetection3d/blob/master/docs/compatibility.md)?

We should.

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (ZwwWayne) [inline mmdet3d/datasets/pipelines/transforms_3d.py:939]: Do we need to keep this function?

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (ZwwWayne) [inline mmdet3d/datasets/pipelines/transforms_3d.py:848]: sample_range might be more clear, indicating the range where the points will be sampled.

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (Wuziyi616) [inline mmdet3d/datasets/pipelines/transforms_3d.py:872]: Would it be better if we use `if sample_range is not None`? After all it's possible value is `int` or `None`. I understand that `if None` is equal to `if False` but I think `not None` is a better style?

File: tests/test_data/test_pipelines/test_indoor_sample.py
Comment (Wuziyi616) [inline tests/test_data/test_pipelines/test_indoor_sample.py:8]: Maybe we should add a case to test `PointSample` with `sample_range is not None`? After all, this is a newly introduced function.

Comment (wHao-Wu) [reply tests/test_data/test_pipelines/test_indoor_sample.py:8]: Thanks for your advice. However, there needs a data conversion from LiDAR to CAMERA or DEPTH before PointSample, which is not supported by the current repo. After a future merge of function 'data conversion', I will add an additional unittest.

Comment (Wuziyi616) [reply tests/test_data/test_pipelines/test_indoor_sample.py:8]: But I guess in unit tests, we can manually construct some Camera points and use them as input? I don't think it's a good idea to 'tie' two transforms together and only test them together? Separately testing makes fewer bugs I believe.

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (wHao-Wu) [reply mmdet3d/datasets/pipelines/transforms_3d.py:872]: Sorry for the late reply. I accidentally missed this review before.

If we sample points from point clouds containing few points, then the sample range will not take effect. And all points, including far points and near points, need to be sampled with the option `replace`.

Comment (wHao-Wu) [reply mmdet3d/datasets/pipelines/transforms_3d.py:872]: Sorry for the late reply. I accidentally missed this review before.

If we sample points from point clouds containing few points, then the sample range will not take effect. And all points, including far points and near points, need to be sampled with the option replace.

File: tests/test_data/test_pipelines/test_indoor_sample.py
Comment (wHao-Wu) [reply tests/test_data/test_pipelines/test_indoor_sample.py:8]: I just added an unittest in `test_transforms_3d.py` to test the point sampling function.

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (Wuziyi616) [reply mmdet3d/datasets/pipelines/transforms_3d.py:872]: Oh, I mean changing it to `if sample_range  is not None and not replace:` . My concern is that `sample_range` is not a `bool` variable, so directly using `if` on it is not standard?

File: tests/test_data/test_pipelines/test_indoor_sample.py
Comment (Wuziyi616) [reply tests/test_data/test_pipelines/test_indoor_sample.py:8]: Yes, great!

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (Wuziyi616) [inline mmdet3d/datasets/pipelines/transforms_3d.py:863]: Also change this to `if sample_range is not None` would be better?

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (Wuziyi616) [inline mmdet3d/datasets/pipelines/transforms_3d.py:894]: Also change to `if self.sample_range is not None` ?

Comment (wHao-Wu) [reply mmdet3d/datasets/pipelines/transforms_3d.py:893]: Comments resolved. Code refined. Thx.

Comment (Wuziyi616) [reply mmdet3d/datasets/pipelines/transforms_3d.py:893]: This line is not resolved yet?

**Review by Wuziyi616 — APPROVED** [review]
Almost ready to merge with one last comment

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (wHao-Wu) [reply mmdet3d/datasets/pipelines/transforms_3d.py:893]: That's just a reminder for developers that we can only support sampling with distance for camera coord system. If necessary, I will switch it to become a `NOTE` message instead of a `TODO` message.

Comment (wHao-Wu) [reply mmdet3d/datasets/pipelines/transforms_3d.py]: Comments resolved. Thx.

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (xiliu8006) [inline mmdet3d/datasets/pipelines/transforms_3d.py:865]: why is the depth = points.coords[:, 2] and not depth=np.linalg.norm(points_numpy[:, 0:3], axis=1) with the points_numpy = points.tensor.clone().numpy()

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (xiliu8006) [inline mmdet3d/datasets/pipelines/transforms_3d.py:850]: why the type of sample_range is int?  if we use it as a threshold to get the far_inds and near_inds, may use the float is better?

Comment (wHao-Wu) [reply mmdet3d/datasets/pipelines/transforms_3d.py:850]: Sorry for that. It's reasonable to switch it to float.

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (wHao-Wu) [reply mmdet3d/datasets/pipelines/transforms_3d.py]: In the original source code of point rcnn, it is true that using the third dimensions to be as the depths of points. I also think it is unreasonable. And the latest code will change the behavior when calculating depth.

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (xiliu8006) [inline mmdet3d/datasets/pipelines/transforms_3d.py:871]: this is necessary？ it is already done in L863

**Review by xiliu8006 — APPROVED** [review]

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (wHao-Wu) [reply mmdet3d/datasets/pipelines/transforms_3d.py:871]: The goals of two `if` statements are different. In the first one, we can determine the sampling range. In the second one, we can combine the sampled near points and far points.

File: mmdet3d/datasets/pipelines/transforms_3d.py
Comment (ZwwWayne) [inline mmdet3d/datasets/pipelines/transforms_3d.py:839]: 1. by default, it should be False? 
2. It seems that it cannot be set from the config/module and it is not used in __call__, so should we keep it?

Comment (wHao-Wu) [reply mmdet3d/datasets/pipelines/transforms_3d.py]: I think `replace` is useful especially when in case that users want to use replace explicitly. So I create an interface for `replace`.

Comment (wHao-Wu) [reply mmdet3d/datasets/pipelines/transforms_3d.py]: Besides, thanks for your remind, I found that most of `replace` in `mmdet3d/datasets/pipelines/transforms_3d.py` are not used in **call**. So there is a choice: keeping shutdown or open for all `replace`.

**Comment by ZwwWayne** [comment]
Can be merged after resolving conflicts

**Pull request merged**
